### DISCUSSION TITLE
[HELP] Mouse events on Textbox do not work well

### DISCUSSION THREAD

**Discussion by IY-science** [fabricjs/fabric.js#9351] [Q&A] [answered]
**disclamier**
I am Japanese and not good at English, so I am using translater and write this.
I trust the translater and make corrections, but I may be getting the wrong message.
If you find something strange, please let me know as you are helping me.

[fabric.js DEMO kitchensink](http://fabricjs.com/kitchensink)
I would like it to go into input waiting mode when I click on the selected Textbox, as shown in the demo at this URL.
My project used to be able to do this, but suddenly it can't.
I can't to decide which is the cause because we have added many functions.

Even after deleting **"canvas.on('mouse:down', () => {})"**, it didn't fix.
I added **"Textbox.on('mousedown', () => {})"** and it responds.
All selection-related properties such as evented, editable, selectable, etc. are true, and all lock-related properties are false.

I use **Rect, Line, Triangle, Ellipse, Image, Textbox and Path** in fabric.Object in my project.
I was experimenting with these objects and found the following.
Note that I consider an **"acrtiveSelection"** that is fabric.Object obtained by **"canvas.getActiveObject()"** and its type is **"activeSelection"**.

- Each of these objects can be selected, group selected, and moved.
- Except for Path, these can't rotate or scale them.
- If those objects are activeSelection, the selections can be rotated, stretched, and collapsed.
- Rotation and scaling can be done by deselecting objects and selecting them individually, regardless of whether or not they are modified after they are activeSelection.
- I can edit a text of Textbox after Textbox becomes an activeSelection at once.

I think the solution is to make it an ActiveSelection, but I would like to be able to perform these operations even if it does not become an ActiveSelection.

I can't solve this problem by myself. I need your help.

https://github.com/fabricjs/fabric.js/assets/115533651/e5493ffa-33af-4337-8dd3-453baa2c15ad

**Comment by asturur** [comment]
@IY-science i moved your translator disclaimer up in the message since it is helpful to know we are reading an automated translated message immedialtely rather than get confused along the way.

Said so, there is no link in your message and indeed is a bit complicated trying to understand what are you trying to achieve.
Could you try to describe the desired outcome instead of what is not working?

Precisely:
```
I would like it to go into input waiting mode when I click on the selected Textbox,
```
Can you elaborate more on this? Even with simple steps like:
- the user click on textbox
- instead of this happening, i would like this other to happen

**Reply by IY-science** [reply]
Thanks for letting me know. And I'm sorry to late response.
I didn't know how to add the URL and didn't do it right. Sorry.
The URL is [fabric.js DEMO kitchensink](http://fabricjs.com/kitchensink).

What I want to do is to enter edit mode when the textbox is selected and clicked.
This is an event that normally fires in Textbox, but for some reason I am not getting it right in my development environment.

The steps are as follows

1. Select the text box.
2. Click the selected text box again.
3. Write text in the text box.

We also want to be able to do the following, as we are no longer able to do so outside of Path.

1. Select objects. 
2. Rotate and scale objects using the displayed controls.

**Reply by asturur** [reply]
ok you broke something in your environment, are you doing a migration from fabric 5 to fabric 6? or developing new code?

**Reply by IY-science** [reply]
The fabric.js's version I use is 5.3.0.

And I created class "Note".

```
class ITextbox extends fabric.Textbox {
  _mouseDownHandler(options: any): void {
      super._mouseDownHandler(options);
  };
};

/* create a new Class "NOTE" */
const note = fabric.util.createClass(fabric.Textbox, {
  type: 'note',
  /*
  initialize: function(options:any){
    options || (options = { });this.callSuper('initialize', options);
  },
  /*
  //initialize cause bad actions
  initialize: function(options:any) {
    options || (options = { });
    this.callSuper('initialize', options);
    this.set('boxStroke', options.boxStroke || '#f00');
    this.set('boxStrokeWidth', options.boxStrokeWidth || 1);
  },
  toObject: function() {
    return fabric.util.object.extend(this.callSuper('toObject'), {
      boxStroke: this.get('boxStroke'),
      boxStrokeWidth: this.get('boxStrokeWidth'),
    });
  },
  */
  _render: function(ctx:any) {
    this.callSuper('_render', ctx);
    ctx.strokeStyle = this.boxStroke;
    ctx.lineWidth = this.strokeWidth
    ctx.strokeRect(this.top, this.left, this.width, this.height)
  },
  /*
  onInput: function(e:any) {
    // Call parent class method
    this.callSuper('onInput', e);
  },
  getIndices: function(subStr:string):number[]{
    const indices:number[] = [];
    let index:number = this.text.indexOf(subStr);
    while(index!=-1){
      indices.push(index);
      index = this.text.indexOf(subStr, index+1);
    }
    return indices;
  }
  */
});


/* add NOTE function
I defined type "otherTextDataType" 
type fontsType = {
    size?: number,
    weight?: boolean,
    style?: boolean,
    underline?: boolean,
    linethrough?: boolean,
    overline?: boolean,
    family?: string,
}
type colorsType = {
    fill?: string,
    stroke?:string,
    textBackgroundColor?: string,
    backgroundColor?: string,
    boxStroke?: string,
}

type otherTextDataType = {
    fonts: fontsType,
    colors: colorsType
}
*/
export const addNote = (text: string, otherTextData: otherTextDataType, width:number, height:number) => {
  const pos = mouseData.staticViewToWorldTransform(userData.sheetState[userData.sheetState.nowLayer].coordinate.nowPos, mouseData.now, userData.sheetState[userData.sheetState.nowLayer].scale);
  const tags:tag[] = [];
  const noteObj: fabric.Textbox = new note(text,{
    maxWidth:width,
    splitByGrapheme: true,
    width, height, 
    top:pos.y,
    left:pos.x,
    fill : otherTextData.colors.fill,
    stroke : otherTextData.colors.stroke,
    backgroundColor : otherTextData.colors.backgroundColor,
    textBackgroundColor : otherTextData.colors.textBackgroundColor,
    boxStroke:otherTextData.colors.boxStroke,
    fontFamily : otherTextData.fonts.family,
    fontSize : otherTextData.fonts.size,
    fontStyle : otherTextData.fonts.style ? 'oblique' : 'normal',
    fontWeight : otherTextData.fonts.weight ? 'bold' : 'normal',
    underline : otherTextData.fonts.underline,
    linethrough : otherTextData.fonts.linethrough,
    overline : otherTextData.fonts.overline,
    editable : true,
    hasControls:true,
    data : {
      beforeText:text,
      class: 'note',
      isEditing:false,
      tags,/*this array has text, uuid, color if any passage selected and painted */
    } ,
  });
  noteObj.on('changed', () => {
    if(noteObj.text==undefined)noteObj.text = '';
    if(noteObj.data.tags.length==0 || noteObj.selectionStart==undefined)return;
    if(noteObj.text.length > noteObj.data.beforeText.length - 1){
      const ss:number = noteObj.selectionStart - 1;
      noteObj.data.tags.forEach( (tag:tag) => {
        if(ss <= tag.start)tag.start++;
        if(ss<=tag.end)tag.end++;
      });
    }
    if(noteObj.text.length < noteObj.data.beforeText.length){
      const ss:number = noteObj.selectionStart + (noteObj.data.beforeText.length - noteObj.text.length);
      noteObj.data.tags.forEach( (tag:tag) => {
        if(ss <= tag.start){tag.start-=ss;tag.end-=ss;}
        else if(ss <= tag.end){tag.end-=ss;}
      });
    }
  })
  addFabric(noteObj);
}

/* this functions support adding fabricObject */ 
export const addFabric = (addData: fabricItems, state?: 'add' | 'preview') => {
  canvas.value.add(addData);
  additionData(addData, {shapeState:state});
  canvas.value.requestRenderAll();
}
export const additionData = (fabricObj: fabricItems, prop:{shapeState?:'add' | 'preview'}) => {
  const uuid = uuidv4();// this function creats uuid that consists of random numbers and English.
  let copiedData:{label:string, uuid:string, timestamp:number, isHidden:boolean};
  copiedData = ('data' in fabricObj) ? {...fabricObj.data} : {};
  //else copiedData = 'data' in (fabricObj.path && fabricObj.path != undefined) ? {...(fabricObj.path as any).data} : {};
  
  copiedData.label = labelData.active;//label name
  copiedData.timestamp = Date.now();//a time that I add
  copiedData.uuid = uuid;
  copiedData.isHidden = false;
  /*
  if(prop.shapeState=='preview'){
    shapeData.preview = uuid;
  }
  sheetData.add.push(uuid);
  if (prop.shapeState!='preview') {
    shapeData.preview = '';
    if(worldData.isOnline)fabricData.created(fabricObj);
  }
  */
  return fabricObj;
}
```

But I tried using the standard Textbox, with the IText extensions turned off, and it behaved strangely as well.

**Comment by IY-science** [comment, marked as answer]
Self-resolved.
It was due to the fact that I was using vue3 and pinia together and had fabric.Canvas in pinia.
I am sorry for the trouble.
